DEVELOPMENT_BEST_PRACTICE.md

```diff
@@ -0,0 +1,308 @@
+
+```go
+func FastenPlanks(tools Toolbox, surface Surface, planks Planks) error {
+    for ; planks.Count() > 0;  {
+        if err := tools.Hammer(surface, planks.Next()); err != nil {
+            return err
+        }
+    }
+}
+```
+
+Avoid constructing dependencies in-line. If you do construct a dependency in-line because it compliments the most common usage, provide a way to override it such as with a setter.
+
+##### "Accept interfaces, return structs"
+
+The Go community talks of the "accept interfaces, return structs" addage. The meaning of this idiom can be hard to pin down, herein lies some clarity.
+
+"Accept interfaces" speaks to the Dependency Inversion Principle (D in 'SOLID'). When a function declares dependent behavior as an interface parameter it gives control of _how_ that behavior is satisfied to the consuming logic. This is the foundation of flexible code.
+
+"return structs" is less obvious. It, generally, speaks to initializing constructors where authors should return the struct being constructed as opposed to interface it satisifes. This is largely due to duck typing (["If it walks like a duck and it quacks like a duck, then it must be a duck"](https://en.wikipedia.org/wiki/Duck_typing)).
+
+> Interfaces in Go provide a way to specify the behavior of an object: if something can do this, then it can be used here. - [Effective Go](https://go.dev/doc/effective_go#interfaces_and_types)
+
+Factory functions are not to be confused with intializing constructors. Factory functions typically return abstractions (an interface) satisfied by 1 of several implementations. A factory could be as simple as:
+
+```go
+// Foo and Bar satisfy fmt.Stringer
+func NewStringer(key string) (fmt.Stringer, error) {
+    switch key {
+    case "foo":
+        return Foo{}, nil
+    case "bar":
+        return Bar{}, nil
+    default:
+        return nil, errors.New("unknown key")
+    }
+}
+```
+
+Factories generally shouldn't contain complex construction logic. That's the job of the initializing constructor for the concrete type.
+
+##### Loggers
+
+Loggers are sometimes treated as a special depency by making entire projects depend on a package level logger. This _masks_ a crucial dependency that is hard to change without impacting the rest of the code base. To understand why this is bad, consider extracting a package into its own repository for reusability. Depending on a package level logger, or any other package level dependencies, makes the extraction process unnecessarily difficult.
+
+Logger's are not special dependencies, inject them like any other dependency. If you want to minimize the surface area of the default initializing constructor (`New...` funcs) you may construct noop loggers to satisfy the logger dependency and provide a means to override the logger with a setter.
```

Comment:
I'm not against this at all. But due to the fact that all the cli code violates this advice, if we want to push for it we should probably pair it with a proposal on how to implement this on new code (eg. inject from the dependency factory to the constructors...) and a plan to little by little refactor existing code.

We should definitely do this separately to this PR but maybe worth adding a note here.

I get this doc aims to help us write better code in general and not only for the CLI. But since that's the piece of software we spent most time with, I figured it might be a good idea to bridge that gap.

Reply:
Definitely. I think the CLI has lots of flaws, actually, that need tidying up - this is 1 of those flaws.

Reply:
I'm making notes on what we need to follow up on and have included this. I'll create a well written issue describing the problem and referencing this guide once its merged.

Reply:
Addressed.